### PR TITLE
A cleaner and more scalable implementation of symbolic tracing

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

This PR provides a much cleaner and less hacky implementation of symbolic tracing for models of the library.
It also provides support for more architectures:

- ALBERT
- DistilBERT
- MobileBERT
- MegatronBERT
- GPT2
- GPT Neo

### CODE REVIEW

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -642,13 +653,28 @@ def _create_and_check_torch_fx_tracing(self, config, inputs_dict, output_loss=Fa
...
+                        continue
+                    else:
+                        flatten.append(x)
+                return flatten
+
+            model_output = flatten_output(model_output)
+            traced_output = flatten_output(traced_output)
             outputs_are_close = all(torch.allclose(model_output[i], traced_output[i]) for i in range(num_outputs))
+
             self.assertTrue(outputs_are_close)
```
Comment (stas00) [inline tests/test_modeling_common.py:678]: 1. I'm thinking that if this ever fails it'd be easier to debug if the loop were to be inverted:
```
for in in range(num_outputs):
    self.assert...
```
would it? I think trying to compare multiple tensors at once is much harder.

If you think it shouldn't really fail, then it is fine as it is.

2. but I'd just add a message to why this may fail, as otherwise it doesn't give context to the failure
```
self.assertTrue(outputs_are_close, "traced output doesn't match model_output")
```

I see that in many of our tests this context message is missing.

**Review by stas00 — APPROVED** [review]
Fantastic re-thinking of how to solve this globally. Great work, @michaelbenayoun!

Added a few minor suggestions.

**Comment by stas00** [comment]
What do you think about `dtype` being hardcoded?

While this is OK for now hardcoding dtype might be an issue down the road. For most NLP models inputs are int, but for example for wav2vec2 they are floats. 

And would this have an impact if the final usage is in fp16 for where you used `float`.

We can't derive the dtype from the model in this context. 

Thoughts?

This is not a showstopper to merge this, but just something to consider - I'm sure we will cross the bridge if we encounter it.

File: src/transformers/modeling_fx_utils.py [hunk 2]
```diff
@@ -21,127 +41,214 @@ def __init__(self, node: Node, tracer: Optional[Tracer] = None):
...
+                inputs_dict["labels"] = torch.ones(batch_size, dtype=torch.long, device=device)
+            elif model_class in get_values(MODEL_FOR_QUESTION_ANSWERING_MAPPING):
+                inputs_dict["start_positions"] = torch.zeros(batch_size, dtype=torch.long, device=device)
+                inputs_dict["end_positions"] = torch.zeros(batch_size, dtype=torch.long, device=device)
+            elif model_class in [
+                *get_values(MODEL_FOR_SEQUENCE_CLASSIFICATION_MAPPING),
+                *get_values(MODEL_FOR_NEXT_SENTENCE_PREDICTION_MAPPING),
+                *get_values(MODEL_FOR_IMAGE_CLASSIFICATION_MAPPING),
+            ]:
+                inputs_dict["labels"] = torch.zeros(batch_size, dtype=torch.long, device=device)
```
Comment (sgugger) [inline src/transformers/modeling_fx_utils.py:187]: This is not correct for sequence classification/image classification models. They can have multiple labels, and those labels could be ints or floats (classification or regression).

**Review by sgugger — APPROVED** [review]
Looks great! I don't think the hardcoding for the dtype of the inputs is a problem since it checks for "id" and "mask" (models that expect float inputs call them something else, like `pixel_values` or `input_values`). However there is a problem in the labels for classification models (both in shapes and dtype).

File: src/transformers/modeling_fx_utils.py (hunk 2 above)
Comment (michaelbenayoun) [reply src/transformers/modeling_fx_utils.py:187]: I will take care of that in the coming PRs!
